### PR TITLE
Ajay - fix person name search on project page frontend

### PR DESCRIPTION
**Pull request by AJAYINAVOLU**
# Description
This PR introduces frontend changes to display a suggestion list when typing a person's name. The suggestions are based on the autocomplete functionality from the backend. When a name is selected from the list, the displayed projects are filtered accordingly, showing only the projects related to the selected person.
Fixes # (bug list priority medium 47)

## Related PRS (if any):
This frontend PR is related to the #[1101](https://github.com/OneCommunityGlobal/HGNRest/pull/1101) backend PR.
To test this fronted PR you need to checkout the #[1101](https://github.com/OneCommunityGlobal/HGNRest/pull/1101) backend PR.
…

## Main changes explained:
- Update **src/actions/userProfile.js** to include actions for fetching person name suggestions and handling selection for project filtering.
- Update **src/components/Projects/Projects.jsx** to implement the logic for filtering projects based on the selected person from the suggestion list.
- Update **src/components/Projects/projects.css** to style the suggestion list and ensure proper display within the project filter UI.
- Update **src/components/SearchProjectByPerson/SearchProjectByPerson.jsx** to add functionality for rendering the suggestion list, handling user input, and triggering project filtering.
- Update **src/constants/userProfile.js** to include constants for managing person-related actions and suggestions.
- Update **src/utils/URL.js** to define API endpoints for fetching person suggestions from the backend.

…

## How to test:
1. Check into the current branch and the #[1101](https://github.com/OneCommunityGlobal/HGNRest/pull/1101) backend PR.
2. Do `npm install` and `npm run start:local` to run this PR locally.
3. Clear site data/cache.
4. Log in as an owner-user.
5. Go to dashboard→ Other Links→ Projects.
6. Go to the Persons's Name search bar and type a name.
7. Verify that the suggestions are shown under the search bar.
8. When the user is selected from the suggestion list, the projects are filtered for only Projects including that person.

## Screenshots or videos of changes:

fr_1
fr_2

https://github.com/user-attachments/assets/ea46f52a-47eb-4570-8732-4a26fed72b5b

### CODE REVIEW

File: src/components/Projects/Projects.jsx [hunk 1]
```diff
@@ -121,6 +123,44 @@ const Projects = function(props) {
     await props.postNewProject(name, category);
   };
 
+  // Fetch autocomplete suggestions
+  useEffect(() => {
+    const fetchSuggestions = async () => {
+      if (debouncedSearchName) {
+        const userSuggestions = await props.getUserByAutocomplete(debouncedSearchName);
+        if (userSuggestions) {
+          setSuggestions(userSuggestions);
```
Comment (navya9989) [inline src/components/Projects/Projects.jsx:132]: I recommend wrapping the fetchSuggestions call in a try-catch block to gracefully handle any errors and provide feedback to the user.

File: src/components/Projects/Projects.jsx [hunk 2]
```diff
@@ -121,6 +123,44 @@ const Projects = function(props) {
     await props.postNewProject(name, category);
   };
 
+  // Fetch autocomplete suggestions
+  useEffect(() => {
+    const fetchSuggestions = async () => {
```
Comment (navya9989) [inline src/components/Projects/Projects.jsx:128]: The fetchSuggestions function is being redefined on every render. While this is usually not a performance issue, you can optimize it by memoizing the function with useCallback if necessary, especially if the component grows in complexity or has performance bottlenecks.
const fetchSuggestions = useCallback(async () => { // Same logic }, [debouncedSearchName, props.getUserByAutocomplete]);

**Review by Sudershhh — CHANGES_REQUESTED** [review]
The Suggestion list seems to end at the very end of the viewport of the screen. 
We can improve this CSS incosistency by limiting it to the end of the search bar for a better UX.
I also like how Navya pointed out the fetchSuggestions issue. If there is a way if you can abstract the function so that it doesnt get rendered on every new render, that would be great! Other than that, looks pretty good to me.

**Review by crystallow1168 — CHANGES_REQUESTED** [review]
Hi @AJAYINAVOLU ,

Below are my findings:

https://github.com/user-attachments/assets/243da411-f9b3-4880-a95a-1bfac57f2e5f

✅ A maximum of 10 suggestions are shown when searching using the search bar
✅ Projects for the selected person is shown when a user is selected from the suggestion list
❗️ The project list does not refresh when the search bar is cleared
❗️ The suggestion list box width to the right is overflowing
❗️ If we have users with the same first and last name, how do we tell them apart. 


<<Hi @crystallow1168, I have fixed the project list refresh and the CSS styling error. Regarding the name matching, this issue is out of the scope of the current fix and would require a design change in the project list layout to distinguish users with the same first and last name. Since we are already pulling the _id of the users in the suggestions, it is possible to differentiate between them if necessary.  

Thanks>> --AJAY

File: src/components/Projects/Projects.jsx (hunk 1 above)
Comment (AJAYINAVOLU) [reply src/components/Projects/Projects.jsx:132]: Fixed it @navya9989.

File: src/components/Projects/Projects.jsx (hunk 2 above)
Comment (AJAYINAVOLU) [reply src/components/Projects/Projects.jsx]: Fixed it @navya9989. Could you please review the PR again?
Thanks

**Review by RishithaMamidala — APPROVED** [review]
@AJAYINAVOLU 
When user is selected from the suggestion list, the projects are filtered for only Projects including that person are returned correctly and on clearing search bar the projects lists also changes correctly.
PR 2676
PR 26756-2

**Review by nikhilpittala16 — APPROVED** [review]
I have tested the PR and everything works as expected.

https://github.com/user-attachments/assets/629490da-56e4-4783-b9c1-27359b44df65

**Review by neeharikakoniki — CHANGES_REQUESTED** [review]
I tested the PR, but unfortunately, the auto-suggestion feature for the person's name doesn't seem to be working as expected. After following the instructions and checking out both the frontend and #1101 backend branches, the suggestions fail to load under the search bar when typing a person's name. Without the auto-suggestions, the main functionality of filtering projects by the selected person cannot be tested effectively. It seems that either the API call for fetching the suggestions isn't functioning correctly, or there’s an issue with how the frontend handles the response. 

image

**Review by Vincent-space — DISMISSED** [review]
Hi, it says Error fetching auto complete.
image

**Review by crystallow1168 — DISMISSED** [review]
Hi @AJAYINAVOLU,
New changes are is working as expected! Great job!

**Review by Saniya98 — APPROVED** [review]
Everything is working as expected

**Comment by one-community** [comment]
Thank you all, merging!

**Pull request merged**